tests/integration_tests/test_advi.py

```diff
@@ -0,0 +1,109 @@
+
+    guide_trace.compute_batch_log_pdf()
+    model_trace.compute_batch_log_pdf()
+
+    assert model_trace.nodes['_advi_latent']['log_pdf'].item() == 0.0
+    assert model_trace.nodes['z1']['log_pdf'].item() != 0.0
+    assert guide_trace.nodes['_advi_latent']['log_pdf'].item() != 0.0
+    assert guide_trace.nodes['z1']['log_pdf'].item() == 0.0
+
+
+# conjugate model to test ADVI logic from end-to-end (this has a non-mean-field posterior)
+class ADVIGaussianChain(GaussianChain):
+
+    # this is gross but we need to convert between different posterior factorizations
+    def compute_target(self, N):
+        self.target_advi_mus = torch.zeros(N + 1)
+        self.target_advi_diag_cov = torch.zeros(N + 1)
+        self.target_advi_mus[-1] = self.target_mus[N].item()
+        self.target_advi_diag_cov[-1] = 1.0 / self.lambda_posts[-1].item()
+        for n in range(N - 1, 0, -1):
+            self.target_advi_mus[n] += self.target_mus[n].item()
+            self.target_advi_mus[n] += self.target_kappas[n].item() * self.target_advi_mus[n + 1]
+            self.target_advi_diag_cov[n] += 1.0 / self.lambda_posts[n].item()
+            self.target_advi_diag_cov[n] += (self.target_kappas[n].item() ** 2) * self.target_advi_diag_cov[n + 1]
+
+    def test_multivariatate_normal_advi(self):
+        self.do_test_advi(3, reparameterized=True, n_steps=10001)
+
+    def do_test_advi(self, N, reparameterized, n_steps):
+        logger.debug("\nGoing to do ADVIGaussianChain test...")
+        pyro.clear_param_store()
+        self.setUp()
+        self.setup_chain(N)
+        self.compute_target(N)
+        self.advi = ADVIMultivariateNormal(self.model)
+        logger.debug("target advi_loc: {}".format(self.target_advi_mus[1:].detach().numpy()))
+        logger.debug("target advi_diag_cov: {}".format(self.target_advi_diag_cov[1:].detach().numpy()))
+
+        # TODO speed up with parallel num_particles > 1
+        adam = optim.Adam({"lr": .0005, "betas": (0.95, 0.999)})
+        svi = SVI(self.advi.model, self.advi.guide, adam, loss="ELBO", trace_graph=False)
+
+        for k in range(n_steps):
+            svi.step(reparameterized)
+
+            if k % 1000 == 0 and k > 0 or k == n_steps - 1:
+                logger.debug("[step {}] advi mean parameter: {}".format(k, pyro.param("advi_loc").detach().numpy()))
+                L = pyro.param("advi_lower_cholesky")
+                diag_cov = torch.mm(L, L.t()).diag()
+                logger.debug("[step {}] advi_diag_cov: {}".format(k, diag_cov.detach().numpy()))
+
+        assert_equal(pyro.param("advi_loc"), self.target_advi_mus[1:], prec=0.05,
+                     msg="advi mean off")
+        assert_equal(diag_cov, self.target_advi_diag_cov[1:], prec=0.07,
+                     msg="advi covariance off")
+
+
+@pytest.mark.parametrize('advi_class', [ADVIDiagonalNormal, ADVIMultivariateNormal])
+def test_diagonal_gaussians(advi_class):
+    n_steps = 3501 if advi_class == ADVIDiagonalNormal else 6001
```

Comment:
tip: another way you could write this is
```py
@pytest.mark.parametrize('advi_class,n_steps', [
    (ADVIDiagonalNormal, 3501),
    (ADVIMultivariateNormal, 6001),
])
def test_diagonal_gaussians(advi_class, n_steps):
    ...
```
which scales a bit better as we add new ADVI classes. no need to change though